src/analyzing/Visit_function_defs.ml

```diff
@@ -130,3 +130,170 @@ let fold_with_class_context
       acc_ref := f !acc_ref opt_ent class_name fdef)
     ast;
   !acc_ref
+
+(* Visitor that tracks both class context and parent function path.
+   The parent_path is a list representing the full path from outermost to innermost:
+   - Top-level function: []
+   - Method: [Some class_name]
+   - Nested function: [None; Some parent_func; Some nested_func] (excluding current function)
+*)
+class ['self] visitor_with_parent_path =
+  object (self : 'self)
+    inherit [_] G.iter_no_id_info as super
+    val current_class : G.name option ref = ref None
+    val parent_path : IL.name option list ref = ref []
+
+    (* Helper to convert G.name to IL.name *)
+    method private g_name_to_il_name (g_name : G.name) : IL.name option =
+      match g_name with
+      | G.Id ((str, tok), id_info) ->
+          let id_info = { id_info with G.id_resolved = ref None } in
+          Some IL.{ ident = (str, tok); sid = G.SId.unsafe_default; id_info }
+      | _ -> None
+
+    (* Helper to get IL.name from entity *)
+    method private entity_to_il_name (ent : G.entity) : IL.name option =
+      match ent.G.name with
+      | G.EN name -> self#g_name_to_il_name name
+      | _ -> None
+
+    method! visit_definition f ((ent, def_kind) as def) =
+      match def_kind with
+      | G.ClassDef _cdef ->
+          let old_class = !current_class in
+          (current_class :=
+             match ent.name with
+             | EN name -> Some name
+             | _ -> None);
+          super#visit_definition f def;
+          current_class := old_class
```

Comment:
There's `Common.save_excursion_unsafe` for this type of pattern. (And `Common.save_excursion` for concurrent use when the ref is in DLS).